### PR TITLE
`seal_reentrant_count`

### PR DESCRIPTION
**Pull request by yarikbratashchuk**
`seal_reentrant_count`, `seal_account_entrance_count`
https://github.com/paritytech/substrate/issues/11018

### CODE REVIEW

File: frame/contracts/src/weights.rs [hunk 1]
```diff
@@ -1748,6 +1761,17 @@ impl WeightInfo for () {
...
+	// Storage: Contracts ContractInfoOf (r:1 w:1)
+	// Storage: Contracts CodeStorage (r:1 w:0)
+	// Storage: Timestamp Now (r:1 w:0)
+	fn seal_reentrant_count(r: u32, ) -> Weight {
+		(0 as Weight)
+		// Standard Error: 2_302_000
+			.saturating_add((922_467_000 as Weight).saturating_mul(r as Weight))
+			.saturating_add(RocksDbWeight::get().reads((99 as Weight).saturating_mul(r as Weight)))
+			.saturating_add(RocksDbWeight::get().writes((99 as Weight).saturating_mul(r as Weight)))
+	}
```
Comment (yarikbratashchuk) [inline frame/contracts/src/weights.rs:1774]: just copy-pasted in order to build, wrong values

File: frame/contracts/src/weights.rs [hunk 2]
```diff
@@ -825,6 +826,18 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
...
+	// Storage: Contracts CodeStorage (r:1 w:0)
+	// Storage: Timestamp Now (r:1 w:0)
+	// Storage: Contracts OwnerInfoOf (r:36 w:36)
+	fn seal_reentrant_count(r: u32, ) -> Weight {
+		(0 as Weight)
+			// Standard Error: 2_302_000
+			.saturating_add((922_467_000 as Weight).saturating_mul(r as Weight))
+			.saturating_add(T::DbWeight::get().reads((99 as Weight).saturating_mul(r as Weight)))
+			.saturating_add(T::DbWeight::get().writes((99 as Weight).saturating_mul(r as Weight)))
+	}
```
Comment (yarikbratashchuk) [inline frame/contracts/src/weights.rs:840]: copy-pasted for build

File: frame/contracts/src/exec.rs [hunk 3]
```diff
@@ -1230,6 +1234,11 @@ where
 		});
 		Ok(())
 	}
+
+	fn reentrant_count(&mut self) -> u32 {
+		let id: &AccountIdOf<Self::T> = &self.top_frame().account_id;
+		(self.frames().filter_map(|f| Some(&f.account_id == id && f.delegate_caller.is_none())).count() - 1).try_into().unwrap()
```
Comment (xgreenx) [inline frame/contracts/src/exec.rs]: Better first do more common check `f.delegate_caller.is_none()` and after `&f.account_id == id`

Comment (xgreenx) [reply frame/contracts/src/exec.rs]: Also maybe the count of frames should affect the weight of the function.

**Comment by xgreenx** [comment]
> Should we pass contract address (account id) into the call, so we don't read it from the memory? @xgreenx

We can do that but better as a separate function to not affect the size of the contract.

**Comment by yarikbratashchuk** [comment]
> > Should we pass contract address (account id) into the call, so we don't read it from the memory? @xgreenx
> 
> We can do that but better as a separate function to not affect the size of the contract.

didn't get it

**Comment by yarikbratashchuk** [comment]
@xgreenx, please take a look

File: frame/contracts/src/wasm/mod.rs [hunk 4]
```diff
@@ -523,6 +525,11 @@ mod tests {
 		fn ecdsa_to_eth_address(&self, _pk: &[u8; 33]) -> Result<[u8; 20], ()> {
 			Ok([2u8; 20])
 		}
+		fn reentrant_count(&mut self) -> u32 {
+			self.reentrant_count.push(12u8);
+			12
+		}
+		fn account_entrance_count(&mut self, _account_id: AccountIdOf<Self::T>) -> u32 { unimplemented!() }
```
Comment (xgreenx) [inline frame/contracts/src/wasm/mod.rs]: Why it is unimplemented?

File: frame/contracts/src/wasm/runtime.rs [hunk 5]
```diff
@@ -2076,4 +2086,20 @@ define_env!(Env, <E: Ext>,
 			Err(_) => Ok(ReturnCode::EcdsaRecoverFailed),
 		}
 	},
+
+	// Returns then number of times currently executing contract exists on the call stack in addition
+	// to the calling instance. A value of 0 means no reentrancy.
+	[__unstable__] seal_reentrant_count(ctx) -> u32 => {
+		ctx.charge_gas(RuntimeCosts::ReentrantCount)?;
+		Ok(ctx.ext.reentrant_count() as u32)
```
Comment (xgreenx) [inline frame/contracts/src/wasm/runtime.rs:2094]: You can use `ctx.ext.account_entrance_count` here and only add charging for reading of the data.

File: frame/contracts/src/exec.rs [hunk 6]
```diff
@@ -1227,6 +1236,15 @@ where
...
 		Ok(())
 	}
+
+	fn reentrant_count(&mut self) -> u32 {
+		let id: AccountIdOf<Self::T> = self.top_frame().account_id.clone();
+		self.account_entrance_count(id) - 1u32
+	}
+
+	fn account_entrance_count(&mut self, account_id: AccountIdOf<Self::T>) -> u32 {
+		self.frames().filter_map(|f| Some(f.delegate_caller.is_none() && &f.account_id == &account_id)).count() as u32
```
Comment (xgreenx) [inline frame/contracts/src/exec.rs]: I guess you can do that in a cheaper way without filtering.

File: frame/contracts/src/exec.rs [hunk 7]
```diff
@@ -1227,6 +1236,15 @@ where
 		});
 		Ok(())
 	}
+
+	fn reentrant_count(&mut self) -> u32 {
+		let id: AccountIdOf<Self::T> = self.top_frame().account_id.clone();
```
Comment (xgreenx) [inline frame/contracts/src/exec.rs]: Why do you need to clone it here? Maybe you can account by ref?

File: frame/contracts/src/benchmarking/mod.rs [hunk 8]
```diff
@@ -2055,6 +2055,59 @@ benchmarks! {
...
+		let r in 0 .. API_BENCHMARK_BATCHES;
+		let code = WasmModule::<T>::from(ModuleDefinition {
+			memory: Some(ImportedMemory::max::<T>()),
+			imported_functions: vec![ImportedFunction {
+				module: "__unstable__",
+				name: "seal_reentrant_count",
+				params: vec![],
+				return_type: Some(ValueType::I32),
+			}],
+			call_body: Some(body::repeated(r * API_BENCHMARK_BATCH_SIZE, &[
```
Comment (xgreenx) [inline frame/contracts/src/benchmarking/mod.rs:2068]: I think the right will be to calculate how much we need gas for one iteration over frames. And in the functions charge exactly for the count of frames. But we can leave it for official review from the `contract-pallet` team if you wish=)

**Review by xgreenx — CHANGES_REQUESTED** [review]

File: frame/contracts/src/wasm/mod.rs (hunk 4 above)
Comment (yarikbratashchuk) [reply frame/contracts/src/wasm/mod.rs]: don't see a good reason to test this

File: frame/contracts/src/wasm/runtime.rs (hunk 5 above)
Comment (yarikbratashchuk) [reply frame/contracts/src/wasm/runtime.rs:2094]: `ctx.ext.reentrant_count` uses `ctx.ext.account_entrance_count`

File: frame/contracts/src/exec.rs (hunk 6 above)
Comment (yarikbratashchuk) [reply frame/contracts/src/exec.rs]: i have no idea why filtering is considered not cheap enough

File: frame/contracts/src/exec.rs (hunk 7 above)
Comment (yarikbratashchuk) [reply frame/contracts/src/exec.rs]: updated

File: frame/contracts/src/benchmarking/mod.rs (hunk 8 above)
Comment (yarikbratashchuk) [reply frame/contracts/src/benchmarking/mod.rs:2068]: Good idea actually. I propose to move forward with it as it is for now.

**Pull request merged**